text_generation/causal_lm/cpp/speculative_decoding_lm.cpp

```diff
@@ -0,0 +1,272 @@
+    and keep the last token predicted by the main network.
+    Then the draft model gets the latest main prediction and again tries to predict 
+    the next K tokens, repeating the cycle.
+    
+    This approach reduces the need for multiple infer requests 
+    to the main model, enhancing performance. For instance, in more predictable 
+    parts of text generation, the draft model can, in best-case scenarios, 
+    generate the next K tokens that exactly match the target. In that case theу
+    are validated in a single inference request of the main model (which is bigger, 
+    more accurate but slower) instead of running K subsequent requests.
+    */
+    int max_sequence_length = 100;
+    while (next_token != SPECIAL_EOS_TOKEN && seq_len < max_sequence_length) {
+        // infer the K next tokens with draft model
+        for (int i = 0; i < K; ++i) {
+            draft_input_ids.data<int64_t>()[0] = next_token;
+            draft_attention_mask.set_shape({BATCH_SIZE, seq_len + i + 1});
+            std::fill_n(draft_attention_mask.data<int64_t>(), draft_attention_mask.get_size(), 1);
+            draft_position_ids.data<int64_t>()[0] = int64_t(draft_attention_mask.get_size() - 1);
+
+            draft_model.infer();
+
+            auto draft_logits = draft_model.get_tensor("logits").data<float>();
+            int64_t arg_max_token = std::max_element(draft_logits, draft_logits + vocab_size) - draft_logits;
+            next_token = arg_max_token;
+            draft_tokens.emplace_back(arg_max_token);
+        }
+
+        // For the main network, K tokens will be fed at once in a single infer request.
+        input_ids.set_shape({BATCH_SIZE, K});
+        // Set the first token for the main model to be the same as for the draft model.
+        input_ids.data<int64_t>()[0] = first_token;
+        for (int i = 0; i < K - 1; i++)
+            input_ids.data<int64_t>()[i + 1] = draft_tokens[i];
+
+        attention_mask.set_shape({BATCH_SIZE, seq_len + K});
+        std::fill_n(attention_mask.data<int64_t>(), attention_mask.get_size(), 1);
+
+        position_ids.set_shape({BATCH_SIZE, K});
+        std::iota(position_ids.data<int64_t>(), position_ids.data<int64_t>() + position_ids.get_size(), seq_len);
+
+        main_model.infer();
+
+        data_logits = logits.data<float>();  // [BATCH_SIZE, K, vocab_size]
+        size_t disagree_idx = K - 1;
+        // Iterate through the predicted tokens from the main model and compare them with draft predictions.
+        // In the worst-case scenario (disagreement at the beginning), iter will increase by 1.
+        // In the best-case scenario, all elements match, and K predicted tokens will be taken.
+        for (size_t i = 0; i < K; i++) {
+            auto start = data_logits + vocab_size * i;
+            auto stop = data_logits + vocab_size * (i + 1);
+            next_token = std::max_element(start, stop) - start;
+            text_streamer.put(next_token);
```

Comment:
should we `put` after `if` below? because current token can be `next_token != draft_tokens[i] `

Reply:
`put` before `if` is the correct way, even if the `next_token != draft_tokens[i]` they both are generated with the same input token and `next_token` contains target model prediction and regardless what is coming next it's always correct. 

If `next_token != draft_tokens[i]` then the token coming after `next_token` will be incorrect. Probably name next is confusing and `out_token` would be better, what do you think?

Reply:
renamed to `out_token`